specifications/transactions/offchain-api/README.md

```diff
@@ -0,0 +1,256 @@
+| code    | str (enum) | Y         | The error code of the corresponding error. |
+| message | str        | N         | Additional details about this error. |
+
+```
+{
+    "type": "protocol_error",
+    "field": "cid",
+    "code": "missing_field",
+    "message": "",
+}
+```
+
+##### Protocol Errors
+
+Use the type `protocol_error` when:
+1. The request has invalid or missing HTTP headers.
+2. A request's `CommandRequestObject`'s fields have validation errors. The CommandRequestObject#command_type must also match the CommandRequestObject#command, but further validation errors are flagged as `command_error`s.
+
+##### Command Errors
+
+Use the type command_error when:
+1. A reqeusts CommandRequestObject#command fields fail to validate.
+2. Unexpected, internal errors.
+
+##### HTTP Responses
+
+* When a validation error occurs, the HTTP response status code should be 400.
+* When a server internal error occurs, HTTP response status code should be 500.
+
+For example, if an off-chain service cannot access its Diem service, it should respond with a 500 status code along with the JWS ecndoed CommandResponseObject.
+
+##### List of Error Codes
+
+The following sections list all error codes for various validations when processing an inbound command request.
+
+##### HTTP Header Validation Error Codes
+
+`invalid_http_header`:
+
+* `X-REQUEST-SENDER-ADDRESS` header value is not the request sender’s address in the command object. All command objects should have a field that is the request sender’s address. For payment objects, it is sender.address or receiver.address.
+* Could not find Diem’s onchain account by the `X-REQUEST-SENDER-ADDRESS` header value.
+* Could not find the compliance key of the onchain account found by the `X-REQUEST-SENDER-ADDRESS` header value.
+* The compliance key found from the onchain account by `X-REQUEST-SENDER-ADDRESS` is not a valid ED25519 public key.
+* `X-REQUEST-ID` is not in valid UUID format.
+
+`missing_http_header`: missing HTTP header `X-REQUEST-ID` or `X-REQUEST-SENDER-ADDRESS`.
+
+####### JWS Validation Error Codes
+
+`invalid_jws`: invalid JWS format (compact) or protected header
+
+`invalid_jws_signature`: JWS signature verification failed
+
+####### Request Object Validation Error Codes
+
+`invalid_json`: decoded JWS body is not json.
+
+`invalid_object`: command request/response object json is not object, or the command object type does not match `command_type`.
+
+`missing_field`: A required field contains no data or is missing.
+
+`unknown_field`: field is unknown for an object.
+
+`unknown_command_type`: invalid/unsupported `command_type`.
+
+`invalid_field_value`: A field does not match the expected type, e.g., a UUID is not correctly formatted or an enum field contains an invalid value.
+
+`invalid_recipient_signature`:
+* Field `recipient_signature` value is not hex-encoded bytes.
+* Field `recipient_signature` value is an invalid signature.
```

Comment:
this is specific to PaymentCommand error, do we want to remove it for now?

Reply:
hmm... probably, let me revisit